File: App/Resources/AlgorithmDetails/inplacemergesort/inplacemergesort.cs
using System;

public class InPlaceMergeSort {
  private static void Push(int[] array, int low, int high) {
    for (var i = low; i < high; i++) {
      if (array[i] > array[i + 1]) {
        (array[i], array[i + 1]) = (array[i + 1], array[i]);
      }
    }
  }

  private static void Merge(int[] array, int low, int high, int mid) {
    var i = low;
    while (i <= mid) {
      if (array[i] > array[mid + 1]) {
        (array[i], array[mid + 1]) = (array[mid + 1], array[i]);
        Push(array, mid + 1, high);
      }
      i++;
    }
  }

  private static void MergeSort(int[] array, int low, int high) {
    if (high - low == 0) {
      return;
    } else if (high - low == 1) {
      if (array[low] > array[high]) {
        (array[low], array[high]) = (array[high], array[low]);
      }
    } else {
      var mid = (low + high) / 2;
      MergeSort(array, low, mid);
      MergeSort(array, mid + 1, high);
      Merge(array, low, high, mid);
    }
  }

  public static int[] Sort(int[] array) {
    if (array.Length >= 2) {
      MergeSort(array, 0, array.Length - 1);
    }
    return array;
  }

  public static void Main(String[] args) {
    int[] array = {0, 39, 21, 62, 91, 77, 14, 23,
      90, 69, 51, 81, 68, 83, 32, 56};
    Sort(array);
    string result = "[" + String.Join(", ", array) + "]";
    Console.WriteLine(result);
  }
}

File: App/Resources/AlgorithmDetails/inplacemergesort/inplacemergesort.rb
def push(array, low, high)
  i = low
  while i < high
    if array[i] > array[i + 1]
      array[i], array[i + 1] = array[i + 1], array[i]
    end
    i += 1
  end
end

def merge(array, low, high, mid)
  i = low
  while i <= mid
    if array[i] > array[mid + 1]
      array[i], array[mid + 1] = array[mid + 1], array[i]
      push(array, mid + 1, high)
    end
    i += 1
  end
end

def merge_sort(array, low, high)
  if high - low == 0
    return
  elsif high - low == 1
    if array[low] > array[high]
      array[low], array[high] = array[high], array[low]
    end
  else
    mid = (low + high) / 2
    merge_sort(array, low, mid)
    merge_sort(array, mid + 1, high)
    merge(array, low, high, mid)
  end
end

def sort(array)
  merge_sort(array, 0, array.length() - 1) if array.length() >= 2
  return array
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array
